train.py

```diff
@@ -230,17 +136,147 @@ def write_bag_cache(bag_file: str, bag_cache_path: str, backbone_onnx_path: str,
+    if offset == 0:
+        if timestamp in entries:
+            return entries[timestamp]
+        else:
+            raise TimedEntryNotFound()
+    elif offset < 0:
+        keys = sorted([key for key in entries.keys() if key < timestamp], reverse=True)
+        offset = abs(offset) - 1
+        if offset > len(keys) - 1:
+            raise TimedEntryNotFound()
+
+        return entries[keys[offset]]
+    else:
+        keys = sorted(key for key in entries.keys() if key > timestamp)
+        if offset > len(keys):
+            raise TimedEntryNotFound()
+        return entries[keys[offset - 1]]
+
+
+@dataclass()
+class DatasetEntry:
+    ts: int
+    observation: np.ndarray
+    observation_yolo_intermediate_size: int
+    action: np.ndarray
+    reward: float
+    done: bool
+
+
+def create_dataset(entries:  Dict[str, Dict[int, np.ndarray]],
+                   backbone_onnx_path: str, reward_func_name: str, interpolation_slice: int) -> List[DatasetEntry]:
+    reward_func = getattr(yolo_reward, reward_func_name)
+    interpolation_time_master = "processed_img"
+    primary = sorted(entries[interpolation_time_master].items())
+    processed_entries = []
+
+    for ts, image_np in primary:
+        bboxes, intermediate = get_onnx_prediction(get_onnx_sess(backbone_onnx_path), image_np)
+        yolo_reward_value = reward_func(bboxes)
+
+        if np.isnan(intermediate).any():
+            raise ValueError("Received NaN in yolo reward calculation")
+
+        yolo_intermediate = _flatten(intermediate)[::interpolation_slice]
 
-    df = pd.DataFrame.from_records([[ts, event, *interps] for (ts, event, interps) in interpolated],
-                                    columns=["ts", *DATAFRAME_COLUMNS], index="ts")
+        try:
+            last_reward_button_connection = get_timed_entry(entries["reward_button_connected"], ts, -1)
+            last_reward_button_override_cmd_vel = get_timed_entry(entries["reward_button_override_cmd_vel"], ts, -1)
+            last_reward_button = get_timed_entry(entries["reward_button"], ts, -1)
+
+            last_head_feedback = get_timed_entry(entries["head_feedback"], ts, -1)
+            last_odrive_feedback = get_timed_entry(entries["odrive_feedback"], ts, -1)
+            last_vbus = get_timed_entry(entries["vbus"], ts, -1)
+            last_head_gyro = get_timed_entry(entries["head_gyro"], ts, -1)
+            last_head_accel = get_timed_entry(entries["head_accel"], ts, -1)
+
+            next_head_cmd = get_timed_entry(entries["head_cmd"], ts, 1)
+            next_cmd_vel = get_timed_entry(entries["cmd_vel"], ts, 1)
+        except TimedEntryNotFound:
+            continue
+
+        if not last_reward_button_connection[0]:
+            print(f"Skipping entry {ts} because reward button app was not connected")
+            continue
+
+        final_reward = yolo_reward_value * opt.base_reward_scale
+
+        move_penalty = abs(next_head_cmd).mean() * 0.002
```

Comment:
This is not really a move penalty. For that we'd need to minimize `abs(next_head_cmd - curr_head_cmd)`.

Although it makes sense to make robot prefer looking straight so maybe keep that penalty too.